src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -781,6 +781,39 @@ cdef CallbackFailureHandler SERVER_SHUTDOWN_FAILURE_HANDLER = CallbackFailureHan
     InternalError)
 
 
+cdef class _ConcurrentRpcLimiter:
+
+    def __cinit__(self, int maximum_concurrent_rpcs, object loop):
+        if maximum_concurrent_rpcs <= 0:
+            raise ValueError("maximum_concurrent_rpcs should be a postive integer")
+        self._maximum_concurrent_rpcs = maximum_concurrent_rpcs
+        self._active_rpcs = 0
+        self._active_rpcs_condition = asyncio.Condition()
+        self._loop = loop
+
+    async def check_before_request_call(self):
+        await self._active_rpcs_condition.acquire()
+        try:
+            predicate = lambda: self._active_rpcs < self._maximum_concurrent_rpcs
+            await self._active_rpcs_condition.wait_for(predicate)
+            self._active_rpcs += 1
+        finally:
+            self._active_rpcs_condition.release()
+
+    def decrease_once_finished(self, object rpc_task):
+        def decrease_active_rpcs_count(_):
+            async def decrease_active_rpcs_count_with_lock():
```

Comment:
I'm not seeing a data dependency from `decrease_active_rpcs_count_with_lock` to `decrease_active_rpcs_count`. Can this function be defined in `decrease_once_finished` instead? I think that will save us from having to create a new function scope each time `decrease_active_rpcs_count` is called.

Reply:
Good idea! Changed.